### PR TITLE
chore(net): integrate discv5 config builder

### PR DESCRIPTION
**Pull request by emhane**
- Integrates `reth_discv5::ConfigBuilder` into `NetworkConfigBuilder`.

The point of having the builder pattern is to build an immutable type. In this case, the config. The way that the network config is currently being built, makes this difficult. This PR tries to keep discv5 config building coherent with discv4 config building, but respecting the builder pattern.
- ~~As many args as possible are added in `DiscoveryArgs::apply_to_builder`, but without passing more other `NetworkArgs` as parameters than necessary.~~ all `DiscoveryArgs` added in outer scope, to enable discv5 by default for optimism
- Bootnodes, a parallel field to `DiscoveryArgs` on `NetworkArgs` type, are added in the outer scope (`NetworkArgs::network_config`).
- Listen sockets are added, wrt to instance, in the outer scope of `NetworkArgs`, in the same function they are added wrt to instance to discv4 (`NetworkArgs::load_network_config`).

At best, an issue is made to design the discv4 config to be immutable too, instead of modifying it after it has been built. ~~Another issue that would make sense, is to move the `instance` flag into `NetworkArgs`.~~ `--instance` is used in `RpcArgs` too, so must exist parallel to `NetworkArgs` and `RpcArgs`.

### CODE REVIEW

File: bin/reth/src/commands/p2p/mod.rs [hunk 1]
```diff
@@ -53,31 +53,14 @@ pub struct Command {
...
-    pub discovery: DiscoveryArgs,
-
-    /// Target trusted peer
-    #[arg(long)]
-    trusted_peer: Option<NodeRecord>,
-
-    /// Connect only to trusted peers
-    #[arg(long)]
-    trusted_only: bool,
+    pub network: NetworkArgs,
```
Comment (mattsse) [inline bin/reth/src/commands/p2p/mod.rs:58]: that makes sense

File: crates/net/common/src/ip.rs
Comment (mattsse) [inline crates/net/common/src/ip.rs:1]: we dont need this, please remove

File: crates/net/network/src/config.rs [hunk 2]
```diff
@@ -434,6 +392,35 @@ impl NetworkConfigBuilder {
         }
     }
 
+    /// Exposes the discovery v5 config builder.
```
Comment (mattsse) [inline crates/net/network/src/config.rs:395]: this should document was the closure actually does, like this gets only called if the builder is set.

File: crates/net/network/src/config.rs [hunk 3]
```diff
@@ -434,6 +392,35 @@ impl NetworkConfigBuilder {
...
+    /// use reth_provider::test_utils::NoopProvider;
+    /// use secp256k1::{rand::thread_rng, SecretKey};
+    ///
+    /// let sk = SecretKey::new(&mut thread_rng());
+    /// let fork_id = MAINNET.latest_fork_id();
+    /// let network_config = NetworkConfigBuilder::new(sk)
+    ///     .discovery_v5_with_builder(|builder| builder.fork(b"eth", fork_id))
+    ///     .build(NoopProvider::default());
+    /// ```
+    pub fn discovery_v5_with_builder(
```
Comment (mattsse) [inline crates/net/network/src/config.rs:408]: consider renaming this function to map or modify

**Review by mattsse — CHANGES_REQUESTED** [review]
does this supersede #7856?

File: crates/net/common/src/ip.rs
Comment (emhane) [reply crates/net/common/src/ip.rs:1]: hm, I'm using it in some places, but will add duplicate code then in, each in the specific files

File: bin/reth/src/commands/p2p/mod.rs [hunk 4]
```diff
@@ -223,3 +209,17 @@ impl Command {
...
 }
+
+fn ipv4(ip: IpAddr) -> Option<Ipv4Addr> {
+    match ip {
+        IpAddr::V4(ip) => Some(ip),
+        IpAddr::V6(_) => None,
+    }
+}
+
+fn ipv6(ip: IpAddr) -> Option<Ipv6Addr> {
```
Comment (mattsse) [inline bin/reth/src/commands/p2p/mod.rs:220]: looks like this is only used once?

can we inline this simple match and remove these functions?

File: crates/node-core/src/node_config.rs [hunk 5]
```diff
@@ -552,16 +533,14 @@ impl Default for NodeConfig {
...
+fn ipv4(ip: IpAddr) -> Option<Ipv4Addr> {
     match ip {
         IpAddr::V4(ip) => Some(ip),
         IpAddr::V6(_) => None,
     }
 }
 
-/// Returns the address if this is an [`Ipv6Addr`].
-pub fn ipv6(ip: IpAddr) -> Option<Ipv6Addr> {
+fn ipv6(ip: IpAddr) -> Option<Ipv6Addr> {
```
Comment (mattsse) [inline crates/node-core/src/node_config.rs:543]: hmm, didn't know this is on main as well, really don't get why IpAddr doesn't have a function for this.

meh the std library does not have this for some reason.

looks like we only use this one, can we remove this function and use match instead?

**Review by mattsse — APPROVED** [review]
pedantic suggestions, otherwise lgtm

**Pull request merged**